README.md

```diff
@@ -511,6 +534,721 @@ will essentially perform a
 [busy-wait](https://en.wikipedia.org/wiki/Busy_waiting), which should usually be
 avoided.
 
+### Programming with explicit transaction log passing
+
+The [`Xt`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Xt/index.html)
+module provides an API that allows transactions to be implemented by explicitly
+passing a mutable transaction log, which allows convenient use of all the
+ordinary sequential control flow structures of OCaml.
+
+> At the moment it is unclear whether both of the
+> [`Xt`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Xt/index.html) and
+> the [`Tx`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Tx/index.html)
+> APIs will be supported in the future. They provide roughly the same expressive
+> power as witnessed by the conversions
+> [`of_tx`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Xt/index.html#val-of_tx)
+> and
+> [`to_tx`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Xt/index.html#val-to_tx).
+> Feedback on this question is welcome!
+
+#### A transactional lock-free leftist heap
+
+Let's implement something a bit more complicated,
+[a leftist heap](https://en.wikipedia.org/wiki/Leftist_tree), which is a kind of
+priority queue.
+
+> The implementation here is adapted from the book _Data Structures and
+> Algorithm Analysis in C (2nd ed.)_ by Mark Allen Weiss.
+
+First we define a data type to represent the spine of a leftist heap:
+
+```ocaml
+# type 'v leftist =
+    [ `Null
+    | `Node of 'v leftist Loc.t
+             * int Loc.t
+             * 'v
+             * 'v leftist Loc.t ]
+type 'v leftist =
+    [ `Node of 'v leftist Loc.t * int Loc.t * 'v * 'v leftist Loc.t | `Null ]
+```
+
+To create a leftist heap we
+[`make`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Loc/index.html#val-make)
+a location with an empty spine:
+
+```ocaml
+# let leftist () : _ leftist Loc.t = Loc.make `Null
+val leftist : unit -> 'a leftist Loc.t = <fun>
+```
+
+We then define an auxiliary function `npl_of` to get the null path length of a
+leftist heap:
+
+```ocaml
+# let npl_of ~xt : _ leftist -> int = function
+    | `Null -> 0
+    | `Node (_, npl, _, _) -> Xt.get ~xt npl
+val npl_of : xt:'a Xt.t -> 'b leftist -> int = <fun>
+```
+
+Notice the `~xt` parameter. It refers to the transaction log being passed
+explicitly. Above we pass it to
+[`get`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Xt/index.html#val-get)
+to record an operation in the log.
+
+The core operation of leftist heaps is that of merging two leftist heaps:
+
+```ocaml
+# let rec merge ~xt ~lt h1 h2 =
+    match h1, h2 with
+    | `Null, h2 -> h2
+    | h1, `Null -> h1
+    | (`Node (_, _, v1, _) as h1),
+      (`Node (_, _, v2, _) as h2) ->
+      let (`Node (h1l, npl, _, h1r) as h1), h2 =
+        if lt v1 v2 then h1, h2 else h2, h1 in
+      let l = Xt.get ~xt h1l in
+      if l == `Null then
+        Xt.set ~xt h1l h2
```

Comment:
It already is.  A leftist tree is left leaning &mdash; If the left branch `l` is `Null`, then the right branch must also be `Null`.  So, in that case the `npl` of the node (with the numbering used in the example) is already 1.

The algorithm is fairly faithfully transliterated from the book mentioned.  One difference is that in the book the npl of `Null` is defined to be `-1`.  In my transliteration I used `0` so the npl values are higher by one.  I had not looked at the wikipedia version before.  The merge operation in the book is written slightly differently (apparently to avoid some swaps and the redundant npl assignment).